### ISSUE TITLE
reattach-to-user-namespace launchctl getenv SSH_AUTH_SOCK does not work

### ISSUE THREAD

**Issue by gzm55** [ChrisJohnsen/tmux-MacOSX-pasteboard#78]
compiling the reattach-to-user-namespace with the latest codes, and test on different term with and without the tmux/screen, `launchctl getenv xxx` still does not work at all.

**Comment by ChrisJohnsen** [comment]
Sorry that you have had difficulties. I can not run Big Sur, so I can only offer limited help.

Certainly you should not use need to use `reattach-to-user-namespace` if *tmux* (or maybe *screen*) is not involved. If your `getenv` is not working in a plain context (i.e. not inside *tmux*/*screen* and not using `reattach-to-user-namespace`), then there is probably not much the wrapper program can help with.

**Comment by gzm55** [comment]
* without tmux, without reattach-to-user-namespace: OK
* without tmux, with reattach-to-user-namespace: NOT OK
* with tmux, without reattach-to-user-namespace: NOT OK
* with tmux, with reattach-to-user-namespace: NOT OK, but should be OK

**Comment by ChrisJohnsen** [comment]
Yeah that last case looks like it is not working as expected. #76 seemed to report that the wrapper was not needed for pasteboard access on Big Sur (I believe this has actually been the case for several macOS releases, not just Big Sur). It could be that whatever Apple did to help support pasteboard access in (e.g.) *tmux* did not also help with [the wrapper’s other uses][bpa].

[bpa]: https://github.com/ChrisJohnsen/tmux-MacOSX-pasteboard/blob/master/Usage.md#beyond-pasteboard-access

Since I can not run Big Sur, I can not really offer any other help at this time. Even if I could, there would be no guarantee of a fix since this whole endeavor has been an exercise in poking around in lightly- and un-documented areas.

I’ll leave the issue open in case anyone else has additional information, suggestions, or patches to try.

**Comment by keith** [comment]
I'm not familiar with what folks use `launchctl getenv` for, so what's a specific repro case for this since I can test on big sur? Just doing this works fine for me in tmux w/ reattach-to-user-namespace:

```
% launchctl setenv BAR test2
% launchctl getenv BAR
test2
```

and any vars I kinda expected to be there don't print in tmux or not:

```
% launchctl getenv HOME
% launchctl getenv USER
% launchctl getenv PATH
```

**Comment by ChrisJohnsen** [comment]
@keith, your `setenv`+`getenv` combination is what I would have used to test it, and its result shows exactly what I would expect to be able to do. Thank you for investigating and testing on Big Sur!

Back here on 10.13 I can peek at all the *launchd*-configured environment variables like this (the manpage warns that the output of the `print` subcommand is not stable across releases, so the information may be in a different place (not inside `environment = { … }`) or this might not work at all in Big Sur):

    launchctl print system | awk '/^\tenvironment = {$/,/^\t}/'    # system-level: usually empty?
    launchctl print user/$UID | awk '/^\tenvironment = {$/,/^\t}/' # user level: usually empty; this is where we “reattach”
    launchctl print gui/$UID | awk '/^\tenvironment = {$/,/^\t}/'  # gui session level: accessible in a Terminal window, but not when “reattached”

---

There must be some complicating factor that prevents `getenv` from working for @gzm55… @gzm55, if the above commands work on Big Sur in which “level” is your variable defined? The particular awk filter might not work if they have restructured the output; you might need to pipe it to (e.g.) *less* and search for your variable’s name or value.

As noted in the comment next to the command, the wrapper only reattaches to the user level, not the GUI level. If your environment variable is only set in the GUI level, then there is no way to get to it with a reattached `launchctl getenv` (which is always at the user level). If you have control over whatever is using `launchctl setenv` to set the variable, then you could try running that from a reattached context (even though it is not inside *tmux*) and that should force it to into the user level where a *tmux*-bound reattached `launchctl getenv` could later “see” it. So instead of `launchctl setenv BAZ quxx` do `reattach-to-user-namespace launchctl setenv BAZ quxx`.

Another possibility is some further conflict between macOS’s session management and *tmux*’s ability to persist (tty) sessions across login sessions. If that is a problem then your `getenv` should work inside a fresh *tmux* server started from a GUI login session (use `tmux -L testing new-session` to start a fresh server and try the `getenv` there). This problem would be harder to solve since the only real hints as to how to deal with macOS’s sessions disappeared when Apple stopped publishing the *launchd* sources back in 10.10.

**Comment by gzm55** [comment]
> launchctl print system | awk '/^\tenvironment = {$/,/^\t}/'    # system-level: usually empty?
> launchctl print user/$UID | awk '/^\tenvironment = {$/,/^\t}/' # user level: usually empty; this is where we “reattach”
> launchctl print gui/$UID | awk '/^\tenvironment = {$/,/^\t}/'  # gui session level: accessible in a Terminal window, but not when “reattached”

what i need is system SSH_AUTH_SOCK env, which is set in gui session level.

**Comment by ChrisJohnsen** [comment]
That variable is set directly by *launchd* (as specified in the *ssh-agent* launch configuration; probably somewhere like `/System/Library/LaunchAgents/com.openssh.ssh-agent.plist`), so you will not be able to control the “level” in which the variable is created. What you might be able to do is copy the value out into a duplicate user-level variable:

    reattach-to-user-namespace launchctl setenv SSH_AUTH_SOCK "$(launchctl getenv SSH_AUTH_SOCK)"

You will need to arrange to run this outside of *tmux*, and it would only need to be run once after each GUI login session is started. You could probably run it during your shell’s startup (i.e. both outside and inside *tmux*) if you add a bit of extra logic to defend against some edge cases:

    if
        s=$(launchctl getenv SSH_AUTH_SOCK 2> /dev/null) && # may print error message when not "attached"
        [ -e "$s" ] && # only use a non-empty value that actually exists as on disk
        u=$(reattach-to-user-namespace launchctl getenv SSH_AUTH_SOCK) &&
        [ "$u" != "$s" ] # only update if different
    then
      reattach-to-user-namespace launchctl setenv SSH_AUTH_SOCK "$s"
    fi
    unset s u

**Comment by gzm55** [comment]
@ChrisJohnsen the copying method in the login session works, thanks for the detail explaination. And in some portable scripts, I could also have to directly parse the env from the `launchctl print` result: 

```bash
 launchctl print gui/$UID | sed -n '/environment/,/}$/ { s/[[:space:]]*SSH_AUTH_SOCK[[:space:]].*[[:space:]]//p; }'
```


Another question, can we attach to the gui session via apple's system api, just like attaching to the user session.

**Comment by ChrisJohnsen** [comment]
Note that, in the *launchctl* manapge (at least my version of it), Apple warns against relying on the structure or content of the output of the `print` subcommand. Extracting the data like that might break without warning in a future release.

Reattaching to the GUI session would be nice, but I have never found a way to do it (there are several promising-looking functions to call, but nothing that I have ever gotten to work; this stuff is all almost completey (publicly) undocumented).

**Comment by gzm55** [comment]
ok, parsing the `launchctl print` message would be the last fall back method with warning, now the stable way should be syncing the value between two session at the shell login process.

**Comment by gzm55** [comment]
Update information:

a protable way to get the ``SSH_AUTH_SOCK`:

```bash
launchctl asuser $UID launchctl getenv SSH_AUTH_SOCK
```

from manual:

>      asuser UID command [args]
>              This executes the given command in as similar an execution context as possible to that of the target user's bootstrap. Adopted attributes include the Mach bootstrap namespace, exception server and security audit session. It does not
>              modify the process' credentials (UID, GID, etc.) or adopt any user-specific environment variables. It affects only the Mach bootstrap context and directly- related attributes.

**Comment by ChrisJohnsen** [comment]
That is a nice find!

I don’t think that subcommand was around back when I started this project, but I do have a vague memory of it requiring root access at some point once it was available. An error message in the code (`asuser_cmd` in [the 10.9.5 version of launchctl.c][lc]) seems to indicate that root was only supposed to be required when requesting a run as a different user, but the code does not seem to check for “different user than current user”, just ”is the current user effectively root” (setting aside the fact that this unprivileged code is an ineffective place for such a permission check). The comment at the top of the function also seems to want to dissuade from the kinds of shenanigans employed therein.

[lc]: https://opensource.apple.com/source/launchd/launchd-842.92.1/support/launchctl.c.auto.html

Ultimately, for releases where it does not require root (at least as early as 10.13), a potentially better version of the wrapper could be this shell script:

    #!/bin/sh
    launchctl asuser "${UID:-"$(id -u)"}" ${1+"$@"}

It could be possible to reverse engineer and recreate the code for modernish versions of `launchctl asuser` (despite the admonition not to), but that is certainly more than I am currently interesting in doing.

**Comment by gzm55** [comment]
i test the code of `asuser_cmd`, but the problem is `_vprocmgr_getsocket`, it always fails with err=46 "(os/kern) service not supported", and the only difference with the current project is two additional environments
```
	setenv(LAUNCHD_SOCKET_ENV, sockpath, 1);
	setenv(LAUNCH_ENV_KEEPCONTEXT, "1", 1);
```

**Comment by ChrisJohnsen** [comment]
I am not too surprised that the old `asuser` code has a few issues on a recent OS release. It looks like those environment variables (`LAUNCHD_SOCKET` and `LaunchKeepContext`, respectively) may no longer be used. I only see `LAUNCHCTL_ENV_REEXEC` added when I do `launchctl asuser $UID env`. It seems like there have been multiple changes in this area as Apple re-writes the *launchd* software. These may or may not be important for a potential reimplementation, but they certainly seem to have a lot or variability between releases.

As long as Apple keeps providing it, I think I will recommend using `asuser` instead of *reattach-to-user-namespace* (unless I find some edge case where it does not work as well).

**Comment by gzm55** [comment]
agree, before `asuser self` is not able to reenter the gui session or require sudo, no one would like to do a reverse engineer in this function.

so this issue could be archived or leave as a reference.

**Comment by glyph** [comment]
For what it's worth this still doesn't seem to work correctly on the latest Monterey (12.3, 21E230) on a notchbook.  (I found this while researching the "managername" quirk of tmux, which is also not fixed with the `reattach-to-user-namespace` in homebrew, either with or without `--HEAD`).

**Comment by gzm55** [comment]
> launchctl asuser

you can use `launchctl asuser ...` as an alternative method.
